File: Searching and Sorting Algorithms/Counting Sort/countingSort.cpp
#include <algorithm>
#include <iostream>
#include <typeinfo>
#include <vector>

using Type = int;

void printArray(std::vector<Type> array) {
  for (const auto& i : array) {
    std::cout << i << " ";
  }
  std::cout << "\n";
}

void countingSort(std::vector<Type>& array) {
  auto max_iter = std::max_element(array.begin(), array.end());
  const int max = array[std::distance(array.begin(), max_iter)];
  std::cout << "Maximum element is " << max << "\n";
  std::vector<Type> output(array.size(), 0);
  std::vector<Type> count(max + 1, 0);
  for (size_t i = 0; i < array.size(); i++)
    count[array[i]] += 1;

  for (size_t i = 1; i < max + 1; i++) {
    count[i] += count[i - 1];
  }

  int i = array.size() - 1;
  while (i >= 0) {
    output[count[array[i]] - 1] = array[i];
    count[array[i]]--;
    i -= 1;
  }

  for (i = 0; i < output.size(); i++) {
    array[i] = output[i];
  }
}

int main() {
  std::vector<Type> array = {6, 4, 9, 4, 3, 5, 1, 14, 7};
  printArray(array);
  countingSort(array);
  printArray(array);
}

File: Searching and Sorting Algorithms/Counting Sort/countingSort.py
def print_array(array):
    for i in array:
        print(i, end=" ")
    print("")


def counting_sort(array):
    maximum = max(array)
    count = [0] * (maximum+1)
    output = [0] * len(array)

    for i in range(len(array)):
        count[array[i]] += 1

    for i in range(1, maximum+1):
        count[i] += count[i-1]

    i = len(array) - 1
    while i >= 0:
        output[count[array[i]] - 1] = array[i]
        count[array[i]] -= 1
        i -= 1

    for i in range(len(output)):
        array[i] = output[i]


foo_array = [9, 4, 3, 5, 1, 8, 2, 7, 0, 6, 3, 2]
print_array(foo_array)
counting_sort(foo_array)
print_array(foo_array)
